build-tools/packages/build-cli/src/library/typescriptApi.ts

```diff
@@ -0,0 +1,97 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type { SourceFile, ExportedDeclarations } from "ts-morph";
+import { Node } from "ts-morph";
+
+import { ApiLevel, isKnownApiLevel } from "./apiLevel.js";
+
+interface ExportRecord {
+	name: string;
+	isTypeOnly: boolean;
+}
+export interface ExportRecords {
+	public: ExportRecord[];
+	beta: ExportRecord[];
+	alpha: ExportRecord[];
+	internal: ExportRecord[];
+	unknown: { name: string; decl: ExportedDeclarations }[];
+}
+
+/**
+ * This function is a placeholder for functionality that could
+ * be very useful to have when working with multiple exports
+ * using the same name (basically both a type and value export).
+ */
+function isTypeExport(_decl: ExportedDeclarations): boolean {
+	return false;
+}
+
+/**
+ * Searches given Node's JSDocs for known {@link ApiLevel} tag.
+ *
+ * @returns Recognized {@link ApiLevel} from JSDocs or undefined.
+ */
+function getNodeLevel(node: Node): ApiLevel | undefined {
+	if (Node.isJSDocable(node)) {
+		for (const jsdoc of node.getJsDocs()) {
+			for (const tag of jsdoc.getTags()) {
+				const tagName = tag.getTagName();
+				if (isKnownApiLevel(tagName)) {
+					return tagName;
+				}
+			}
+		}
+	} else {
+		// Some nodes like `VariableDeclaration`s as not JSDocable, but an ancestor
+		// like `VariableStatement` is and may contain tag.
+		const parent = node.getParent();
+		if (parent !== undefined) {
+			return getNodeLevel(parent);
+		}
+	}
+	return undefined;
+}
+
+/**
+ * Given a source file extracts all of the named exports and associated API level.
+ * Named exports without a recognized level are placed in unknown array.
+ */
+export function getApiExports(sourceFile: SourceFile): ExportRecords {
+	const exported = sourceFile.getExportedDeclarations();
+	const records: ExportRecords = {
+		public: [],
+		beta: [],
+		alpha: [],
+		internal: [],
+		unknown: [],
```

Comment:
Unknown here means untagged? Presumably this shouldn't happen (we are configured to throw errors on package exports with missing tags). So we could fail if we see them.

Reply:
Which I see that we do, just elsewhere (aggregated). Looks good!

Reply:
Yes. This helper will be called by two cases. I don't expect them in either case, but the one should be more forgiving as it will run against packages we have shipped and we won't have easy control to change.
This code was refactored from fluid-imports.ts, but I haven't removed the duplicate code from there. The unknown is a new feature in this refactor.
The `generate api` will log an error for such cases. I stopped short of throwing an exception. @tylerbutler, `logError` won't cause process to fail, right?